### PR TITLE
cilium-cni: Call IPAMAllocate before creating veth

### LINKED ISSUE #19004: Cilium keeps creating temporary network interfaces when it can't allocate more IPs
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

i installed cilium on eks with eni mode. when cilium cannot allocate more IPs, pods fails to start (which is expected):
```
  Warning  FailedCreatePodSandBox  9m57s                    kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to set up sandbox container "b217c5d39aa90120cc5c7d814dc3747b95ecbe597fc7ae38833e1d4481bc8e61" network for pod "nginx-deployment-66b6c48dd5-d6lbr": networkPlugin cni failed to set up pod "nginx-deployment-66b6c48dd5-d6lbr_default" network: unable to allocate IP via local cilium agent: [POST /ipam][502] postIpamFailure  No more IPs available
```
when cilium fails to initialize the pods, kubernetes keeps calling cilium-cni, and each call creates a new temporary interface.
```
% dmesg -wT
[Wed Mar  2 04:24:10 2022] eth0: renamed from tmp2c88e
[Wed Mar  2 04:24:11 2022] eth0: renamed from tmpeea9f
[Wed Mar  2 04:24:13 2022] eth0: renamed from tmp57161
[Wed Mar  2 04:24:14 2022] eth0: renamed from tmp0201c
...
```
at some point some pods get rescheduled to some other nodes, and cilium succeeds to allocate a new IP. then cilium crashes because the interface index is now too high:
```
level=fatal msg="ens6 link ifindex 317689 exceeds max(uint16)" subsys=daemon 
```
assuming an interface gets created every second, it takes 65535 seconds (about 18 hours) to hit the limit. restarting the node fixes the issue.

### Cilium Version

```
% cilium version
cilium-cli: v0.10.4 compiled with go1.17.7 on linux/amd64
cilium image (default): v1.11.2
cilium image (stable): v1.11.2
cilium image (running): v1.11.2
```

### Kernel Version

`5.4.176-91.338.amzn2.x86_64`

### Kubernetes Version

```
% kubectl version
Client Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.4", GitCommit:"e6c093d87ea4cbb530a7b2ae91e54c0842d8308a", GitTreeState:"clean", BuildDate:"2022-02-16T12:38:05Z", GoVersion:"go1.17.7", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"21+", GitVersion:"v1.21.5-eks-bc4871b", GitCommit:"5236faf39f1b7a7dabea8df12726f25608131aa9", GitTreeState:"clean", BuildDate:"2021-10-29T23:32:16Z", GoVersion:"go1.16.8", Compiler:"gc", Platform:"linux/amd64"}
```

### Sysdump

i'll share it privately if necessary.

### Relevant log output

_No response_

### Anything else?

i opened a draft PR: https://github.com/cilium/cilium/pull/18975 the idea here is to try calling `c.IPAMAllocate()` before creating an interface so that we don't keep creating temporary interfaces when `c.IPAMAllocate()` fails.

steps to repro:
```
% export NAME="$(whoami)-$RANDOM"

% cat <<EOF >eks-config.yaml
apiVersion: eksctl.io/v1alpha5
kind: ClusterConfig

metadata:
  name: ${NAME}
  region: eu-west-1

managedNodeGroups:
- name: ng-1
  desiredCapacity: 2
  privateNetworking: true
  # taint nodes so that application pods are
  # not scheduled until Cilium is deployed.
  taints:
   - key: "node.cilium.io/agent-not-ready"
     value: "true"
     effect: "NoSchedule"
EOF

% eksctl create cluster -f ./eks-config.yaml

% cilium install

% cilium version
cilium-cli: v0.10.4 compiled with go1.17.7 on linux/amd64
cilium image (default): v1.11.2
cilium image (stable): v1.11.2
cilium image (running): v1.11.2

% cat deployment.yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx-deployment
  labels:
    app: nginx
spec:
  replicas: 50
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - name: nginx
        image: nginx:1.14.2
        ports:
        - containerPort: 80

% kubectl apply -f ./deployment.yaml
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by michi-covalent**
Call IPAMAllocate before creating a temporary network interface to avoid
unnecessarily incrementing the network interface index. Without this
patch, cilium-cni keeps creating temporary interfaces when it cannot
allocate any more IPs. Eventually it exceeds max(uint16) and Cilium
crashes with an error like this:

    level=fatal msg="ens6 link ifindex 317689 exceeds max(uint16)" subsys=daemon

Fixes: #19004

Signed-off-by: Michi Mutsuzaki <michi@isovalent.com>

```release-note
Fix a bug where Cilium would constantly create network interfaces if IPAM limits are reached
```

### CODE REVIEW

**Comment by joestringer** [comment]
Suggested release-note: `Fix bug where Cilium would constantly create network interfaces if IPAM limits are reached"

File: plugins/cilium-cni/cilium-cni.go [hunk 1]
```diff
@@ -389,6 +389,26 @@ func cmdAdd(args *skel.CmdArgs) (err error) {
...
+	}
+
+	// release addresses on failure
+	defer func() {
+		if err != nil && ipam != nil && ipam.Address != nil {
+			releaseIP(c, ipam.Address.IPV4)
+			releaseIP(c, ipam.Address.IPV6)
+		}
+	}()
+
```
Comment (sayboras) [inline plugins/cilium-cni/cilium-cni.go:411]: should we move the below validation earlier as well ?

```
	if err = connector.SufficientAddressing(ipam.HostAddressing); err != nil {
		err = fmt.Errorf("IP allocation addressing in insufficient: %s", err)
		return
	}
```

**Review by sayboras — APPROVED** [review]

File: plugins/cilium-cni/cilium-cni.go (hunk 1 above)
Comment (michi-covalent) [reply plugins/cilium-cni/cilium-cni.go:411]: yeah looks like we can move that block as well. let me try it.

Comment (michi-covalent) [reply plugins/cilium-cni/cilium-cni.go:411]: updated. running the test now :+1:

**Review by joestringer — APPROVED** [review]

File: plugins/cilium-cni/cilium-cni.go (hunk 1 above)
Comment (michi-covalent) [reply plugins/cilium-cni/cilium-cni.go:411]: :heavy_check_mark: ran the same test as https://github.com/cilium/cilium/pull/18975#issuecomment-1056253189. working well.

**Review by pchaigno — APPROVED** [review]

**Comment by michi-covalent** [comment]
slightly confused about the test failure. according to https://jenkins.cilium.io/job/Cilium-PR-K8s-GKE/7768/testReport/junit/Suite-k8s-1/20/K8sHealthTest_cilium_health_Checks_status_between_nodes/
```
Stderr:
 	 Error from server (NotFound): deployments.apps "coredns" not found
```
but looks like the deployment name is `kube-dns`, not `coredns` :thinking: 
```
kube-system         kube-dns                        1/2     2            1           9h      kubedns,dnsmasq,sidecar,prometheus-to-sd   gke.gcr.io/k8s-dns-kube-dns:1.21.0-gke.0,gke.gcr.io/k8s-dns-dnsmasq-nanny:1.21.0-gke.0,gke.gcr.io/k8s-dns-sidecar:1.21.0-gke.0,gke.gcr.io/prometheus-to-sd:v0.4.2   k8s-app=kube-dns
```

**Comment by michi-covalent** [comment]
... never mind, looks like it checks both: https://github.com/cilium/cilium/blob/83d74a255487acaad9caad44a04c1ea5f8ef8cb6/test/helpers/kubectl.go#L1977. looks like kube-dns had trouble connecting to k8s api server.
```
2022-03-02T19:37:14.538206894Z I0302 19:37:14.537994       1 dns.go:221] Waiting for [endpoints services] to be initialized from apiserver...
2022-03-02T19:37:14.538732794Z I0302 19:37:14.538231       1 trace.go:205] Trace[1298498081]: "Reflector ListAndWatch" name:pkg/mod/k8s.io/client-go@v0.19.12/tools/cache/reflector.go:156 (02-Mar-2022 19:36:44.537) (total time: 30000ms):
2022-03-02T19:37:14.538768699Z Trace[1298498081]: [30.000937368s] [30.000937368s] END
2022-03-02T19:37:14.538785086Z E0302 19:37:14.538276       1 reflector.go:127] pkg/mod/k8s.io/client-go@v0.19.12/tools/cache/reflector.go:156: Failed to watch *v1.Service: failed to list *v1.Service: Get "https://10.35.240.1:443/api/v1/services?limit=500&resourceVersion=0": dial tcp 10.35.240.1:443: i/o timeout
2022-03-02T19:37:14.538793764Z I0302 19:37:14.538307       1 trace.go:205] Trace[2019727887]: "Reflector ListAndWatch" name:pkg/mod/k8s.io/client-go@v0.19.12/tools/cache/reflector.go:156 (02-Mar-2022 19:36:44.537) (total time: 30000ms):
2022-03-02T19:37:14.538797870Z Trace[2019727887]: [30.00089245s] [30.00089245s] END
2022-03-02T19:37:14.538802181Z E0302 19:37:14.538337       1 reflector.go:127] pkg/mod/k8s.io/client-go@v0.19.12/tools/cache/reflector.go:156: Failed to watch *v1.Endpoints: failed to list *v1.Endpoints: Get "https://10.35.240.1:443/api/v1/endpoints?limit=500&resourceVersion=0": dial tcp 10.35.240.1:443: i/o timeout
```
this could just be a transient issue. re-triggering the test.

**Pull request merged**